### PR TITLE
Participation: count atts from orphaned blocks

### LINKED ISSUE #7116: eth/v1alpha1/validators/participation returns wrong data
> This API only gives you the voted ether precisely at the end of epoch `N` whereas it can still grow through` N+1`
> So at the end of epoch 4805, you can run that endpoint for 4805 and get the balances then at the end of 4806, you can look at "previous" and get the final balance result
> 
> Is the participation rate the head, target, or source pariticpation?
> "target" is what matters for justification so it might be the relevant number to show
> If there was a bunch of forking, you could imagine a source participation that was much higher than target and thus depending n what was displayed could be confusing

http://localhost:3500/eth/v1alpha1/validators/participation?epoch=4805
```
epoch    "4805"
finalized    true
participation    
globalParticipationRate    0.36303958
votedEther    "373192000000000"
eligibleEther    "1027965000000000"
```
Lighthouse reports 78%
https://lighthouse-book.sigmaprime.io/http/consensus.html#consensusglobal_votes

quoting @djrtwo here as he found the issue, thanks :)

### PR DESCRIPTION
**Pull request by terencechain**
**What type of PR is this?**
> Bug fix

**What does this PR do? Why is it needed?**
Participation end point had a bug, it excludes attestations from orphaned block. This PR fixed it by appending orphaned block's attestations back into requested state and also using a more reasonable end slot

**Which issues(s) does this PR fix?**

Fixes #7116

**Other notes for review**

### CODE REVIEW

File: beacon-chain/rpc/beacon/validators.go [hunk 1]
```diff
@@ -475,20 +477,22 @@ func (bs *Server) GetValidatorParticipation(
 		)
 	}
 
-	requestedState, err := bs.StateGen.StateBySlot(ctx, helpers.StartSlot(requestedEpoch+1))
+	nextEpochEndSlot := helpers.StartSlot(requestedEpoch+2) - 1
```
Comment (nisdas) [inline beacon-chain/rpc/beacon/validators.go:480]: what are these magic numbers (2 & 1)  ? Maybe we can declare them as vars here.

File: beacon-chain/rpc/beacon/validators.go [hunk 2]
```diff
@@ -505,6 +509,48 @@ func (bs *Server) GetValidatorParticipation(
...
+// nor finalized blocks. To count participation, this includes the attestations from the orphaned block to state.
+func (bs *Server) appendNonFinalizedBlockAttsToState(ctx context.Context, s *statetrie.BeaconState, e uint64) (*statetrie.BeaconState, error) {
+	start := helpers.StartSlot(e)
+	end := helpers.StartSlot(e + 1)
+	f := filters.NewFilter().SetStartSlot(start).SetEndSlot(end)
+	blks, err := bs.BeaconDB.Blocks(ctx, f)
+	if err != nil {
+		return nil, err
+	}
+	bRoots, err := bs.BeaconDB.BlockRoots(ctx, f)
```
Comment (nisdas) [inline beacon-chain/rpc/beacon/validators.go:522]: fyi , there is an issue with using this method. During the medalla incident there were times when the block roots did not match up to the blocks retrieved. #7028

Might be better to just hash them here.

File: beacon-chain/rpc/beacon/validators.go [hunk 3]
```diff
@@ -505,6 +509,48 @@ func (bs *Server) GetValidatorParticipation(
...
+	if err != nil {
+		return nil, err
+	}
+	if len(blks) != len(bRoots) {
+		return nil, err
+	}
+	fbs := make([]*ethpb.SignedBeaconBlock, 0, len(blks))
+	for i := len(blks) - 1; i >= 0; i-- {
```
Comment (nisdas) [inline beacon-chain/rpc/beacon/validators.go:530]: what is the point of this, this slice isn't used anywhere after

File: beacon-chain/rpc/beacon/validators.go (hunk 1 above)
Comment (terencechain) [reply beacon-chain/rpc/beacon/validators.go:480]: It's to get `nextEpochEndSlot`, I think the var name is good enough. Forward two epochs and minus one

File: beacon-chain/rpc/beacon/validators.go (hunk 2 above)
Comment (terencechain) [reply beacon-chain/rpc/beacon/validators.go]: Will do

File: beacon-chain/rpc/beacon/validators.go (hunk 3 above)
Comment (terencechain) [reply beacon-chain/rpc/beacon/validators.go:526]: Nice catch. Fixed this in the latest commit. Thanks!

**Review by nisdas — DISMISSED** [review]
lgtm :+1:

**Review by nisdas — APPROVED** [review]

**Pull request merged**